Fix/mAP

rafaelpadilla opened the pull request:
# Description

The current implementation of **Mean Average Precision (mAP)** in `supervision.metrics` produces results that diverge from both **pycocotools** and published benchmarks (e.g., the [Roboflow leaderboard](https://leaderboard.roboflow.com/)).

This PR **aligns supervision's mAP implementation with pycocotools**, the official COCO evaluation tool, ensuring reliable, standardized metrics. 🚀

## Key points 🏆
* **Self-contained**: No external dependencies are needed to compute the metric. 
* **Seamless integration**: Metrics are computed directly in `supervision/metrics/mean_average_precision.py` and works directly with `supervision.Detections` objects.
* **Unchanged public API**: Existing user code continues to work with the same interface. **No impact for users!** 🔥
```
import supervision as sv
from supervision.metrics import MeanAveragePrecision
preds   = sv.Detections(...)
targets = sv.Detections(...)
metric  = MeanAveragePrecision()
result  = metric.update(preds, targets).compute()
print(result.map50_95)

```

* **Same result object**: `MeanAveragePrecisionResult` class still exposes the same properties:
```
print("Overall:")
print(f"map50_95: {result.map50_95}")
print(f"map50:     {result.map50}")
print(f"map75:     {result.map75}")
print(f"map50_95: {result.small_objects.map50_95}")
print(f"map50_95: {result.medium_objects.map50_95}")
print(f"map50_95: {result.large_objects.map50_95}")
```
* **Numerical parity**: Evaluations match pycocotools within < 2e-5. ✅
    * A Colab notebook was created to compute mAP results for RT-Detr models **purely with pycocotools** [here](https://colab.research.google.com/drive/1seSGLmmJvQJ4JbXkjt_nY32UuLUXayRa?usp=sharing).
    * Another Colab notebook was created to compute mAP results for RT-Detr models using **this proposed branch** [here](https://colab.research.google.com/drive/1fFjMC9_K4z4VTYC7mFlsmIR4YyV_Dhgx?usp=sharing).
    * A detailed results table comparing the results of both Colab notebooks (pure pycocotools vs. this PR) is [here](https://docs.google.com/spreadsheets/d/1w8MpCgJuv6vZQfNTbmzNgitKvJwynuLpB9Ioqxkm-SI/edit?gid=0), showing a maximum deviation of 2e-5. 🎯

## Type of change
-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

SkalskiP commented on supervision/metrics/mean_average_precision.py:
```diff
@@ -626,3 +229,1253 @@ def plot(self):
         plt.tight_layout()
         plt.show()
+
+    def to_pycocotools_output(self) -> str:
```
`MeanAveragePrecisionResult` currently have `__str__`. I think we can replace `__str__` implementation with this.

rafaelpadilla replied:
Done!

SkalskiP commented on supervision/metrics/mean_average_precision.py:
```diff
@@ -626,3 +229,1253 @@ def plot(self):
+        """
+        Initialize the Mean Average Precision metric.
+
+        Args:
```
Update docstring to reflect newly added arguments.

rafaelpadilla replied:
My bad. Thanks for noticing it :smile:

SkalskiP commented on supervision/metrics/mean_average_precision.py:
```diff
@@ -626,3 +229,1253 @@ def plot(self):
+
+        return self
+
+    def _compute_polygon_area(self, coords: List[float]) -> float:
```
Looks like this function is unused.

rafaelpadilla replied:
Removed!

SkalskiP commented on supervision/dataset/formats/coco.py:
```diff
@@ -102,15 +106,35 @@ def coco_annotations_to_detections(
+    else:
+        area = None
+
+    if use_iscrowd or use_precomputed_area:
```
Does it make sense to have two separate flags if, in the end, they are connected with `or` ?

rafaelpadilla replied:
Thanks for bringing this up — you're absolutely right. Initially, I aimed to keep each part independent to provide more flexibility for external users. However, after reviewing it more closely, I agree that a single flag is sufficient and cleaner. I've updated the code accordingly to keep only one flag.

rafaelpadilla replied:
I agree that `use_iscrowd` might be a bit misleading when you first see it. However, the evaluation logic needs to account for the `iscrowd` property when it’s present in the dataset. This is necessary to produce results consistent with pycocotools.

Also, the `use_iscrowd` flag (which defaults to `False`) is essential for maintaining compatibility with existing tests, particularly test/dataset/formats/test_coco.py::test_coco_annotations_to_detections.

In short, the `use_iscrowd` flag serves two purposes:
* It includes the `iscrowd` property from the dataset in the Detection objects, which is critical for matching pycocotools behavior.
* It ensures backward compatibility with existing tests.

To keep things simple and closer to the original logic, I kept the `use_iscrowd` flag only in the `coco_annotations_to_detections` function. I hope that addresses your concern, but happy to adjust if you have other suggestions.

SkalskiP commented on supervision/dataset/formats/coco.py:
```diff
@@ -159,16 +183,29 @@ def detections_to_coco_annotations(
     return coco_annotations, annotation_id
 
 
+def get_coco_class_index_mapping(annotations_path: str) -> Dict[int, int]:
```
This function is not used.

rafaelpadilla replied:
You're correct — this function is not currently invoked in the main pipeline. However, it is essential for evaluating certain models whose class ID schemes differ from those used in the COCO dataset.

Specifically, some models use sequential class IDs (e.g., 0 to 79 for 80 classes), whereas COCO's official annotations intentionally skip some IDs. You can see a detailed breakdown of these skipped IDs in [this spread sheet](https://docs.google.com/spreadsheets/d/e/2PACX-1vTTq7tjBi4RXPKS4_mCXEM4PF6i3YpRG6X1hvJjS_7OQjN4-D6DEBUg8OzGDat9fklk9iF3zdxrWDNK/pubhtml).

To address this mismatch, this function is super useful. A practical example of this mapping is used in
[this colab notebook](https://colab.research.google.com/drive/1fFjMC9_K4z4VTYC7mFlsmIR4YyV_Dhgx?usp=sharing), where `get_coco_class_index_mapping` is applied to reproduce results consistent with the [roboflow/model-leaderboard](https://github.com/roboflow/model-leaderboard)

rafaelpadilla replied:
I have updated the code to return the reversed mapping. Thanks for the suggestion!

rafaelpadilla commented on supervision/metrics/mean_average_precision.py:
```diff
@@ -626,3 +229,1253 @@ def plot(self):
+        )
+
+
+class EvaluationDataset:
```
The idea of having an intermediate format provided by the `EvaluationDataset` object is that we keep the implementation of `COCOEvaluator` as close as possible to pycocotools. 

`Detections` objects do not have properties that facilitate the search of objects and images (like `image id`) in a large dataset. The `EvaluationDataset` class facilitates the process, as it contains functions to find detections and images easily, just like in pycocotools.

In the future, if the object `Detections` is changed so that it can differentiate "predictions" from "annotations", and includes image ids, I think it would make sense to change.

The `EvaluationDataset` facilitates maintainability, comprehension, comparability with pycocotools (super important imho), and facilitates the expansion of the code to other metrics like average recall.

rafaelpadilla replied:
The main benefit of introducing an `EvaluationDataset` intermediate is alignment with `pycocotools`. It mirrors their internal structure and makes our implementation easier to understand, debug, and compare against their results — which is crucial for metric consistency and correctness.

Also, `Detections` doesn't currently expose image IDs or a clean way to separate predictions from ground truth. `EvaluationDataset` fills that gap by organizing data in a way that's much closer to what `pycocotools` expects, making it easier to implement metrics like mAP and potentially recall in the future.

If the `Detections` class evolves to include this structure natively, we could definitely revisit. But for now, this abstraction helps keep the logic cleaner, more testable, and easier to extend without bloating the evaluator.

rafaelpadilla commented on supervision/metrics/mean_average_precision.py:
```diff
@@ -626,3 +229,1253 @@ def plot(self):
+    if is_crowd:
+        return Ai / (Aa + EPS)
+
+    return Ai / (Aa + Ab - Ai + EPS)
```
Done!

rafaelpadilla commented:
Hey @SkalskiP ,

Thank you for your review! 🙌
I've addressed all the comments and updated the code accordingly.
Let me know if everything looks good on your side.

SkalskiP commented on supervision/dataset/formats/coco.py:
```diff
@@ -102,15 +105,29 @@ def coco_annotations_to_detections(
+            iscrowd=np.asarray(iscrowd, dtype=int), area=np.asarray(area, dtype=float)
+        )
+    else:
+        iscrowd = [0] * len(image_annotations)
+        area = None
```
It looks like we’re not actually using `iscrowd` or `area`; we’re just returning empty dictionaries. I think it makes sense to remove both `iscrowd` and `area`.

SkalskiP commented on supervision/metrics/mean_average_precision.py:
```diff
@@ -626,3 +208,1144 @@ def plot(self):
+        dt_boxes = [d["bbox"] for d in dt]
+
+        # Get the iscrowd flag for each gt
+        is_crowd = [int(o["iscrowd"]) for o in gt]
```
When you load a dataset that doesn’t have `iscrowd`, this line causes the entire code to crash. We want users to be able to run evaluation on their own datasets, which might be in YOLO format or even COCO format without `iscrowd`. The code shouldn’t crash in this case.

rafaelpadilla replied:
You're absolutely right — thanks for pointing that out! 
:heavy_check_mark:  I've updated the evaluation logic so that if `iscrowd` is missing from the dataset, it now defaults to `iscrowd=0`. 
This ensures compatibility with datasets in YOLO format or COCO variants that don’t include the `iscrowd` field, and prevents the code from crashing in those cases.

SkalskiP commented:
Hi @rafaelpadilla, I just tested YOLO11 by comparing the results obtained with the updated `supervision` to those from `pycocotools`, and we’re seeing about a ~1% discrepancy.

SkalskiP commented:
Hi @rafaelpadilla, short update: I spent some time today diving a bit deeper. I independently benchmarked YOLO11 using both the new `supervision` and `pycocotools`, and I’m getting identical results between those two methods. However, these results differ from what our ML team report in the paper. I reached out to our ML team, to learn more.

Here are Google Colabs I prepared:

- [YOLO11 benchmark with fixed supervision](https://colab.research.google.com/drive/1ZPsO6d0R0Btqs2iApyvr-ieAeWejBHco?usp=sharing)
- [YOLO11 benchmark with pycocotools](https://colab.research.google.com/drive/13hrbP7P302l0qmHxhKxfO9T9m6A-FH8J?usp=sharing)

I'll keep you posted. For now, let's focus on other comments in this PR.

rafaelpadilla commented:
Hey Piotr,

Thanks again for your review!

I've addressed all your comments and incorporated your suggestions throughout the PR.

Regarding results consistency:
* I re-ran the tests using your YOLO Colab [notebook](https://colab.research.google.com/drive/1ZPsO6d0R0Btqs2iApyvr-ieAeWejBHco?usp=sharing) and confirmed that the results remain identical :rocket: 
* I also re-tested with RT-DETR using my [notebook](https://colab.research.google.com/drive/1fFjMC9_K4z4VTYC7mFlsmIR4YyV_Dhgx?usp=sharing#scrollTo=_tL9JuQeF-Nu)  and the results are consistent as well. :rocket: 
* To make things easier to compare, I’ve consolidated both sets of results (YOLO and RT-DETR) into [this Google Sheet](https://docs.google.com/spreadsheets/d/1w8MpCgJuv6vZQfNTbmzNgitKvJwynuLpB9Ioqxkm-SI/edit?usp=sharing).

All points that could be implemented without impacting metric accuracy have been addressed. For the others, I provided context explaining the rationale behind the current implementation and pycocotools.

I believe this PR will make Supervision a simple and reliable tool for testing object detection models.

Let me know if anything else comes up.

rafaelpadilla commented:
Sorry, I think I accidentally closed this PR. 
Reopening it

SkalskiP commented on supervision/dataset/formats/coco.py:
```diff
@@ -159,16 +173,29 @@ def detections_to_coco_annotations(
     images_directory_path: str,
     annotations_path: str,
     force_masks: bool = False,
+    # use_iscrowd: bool = True,
```
I noticed that in [this commit](https://github.com/roboflow/supervision/pull/1834/commits/cff974555e9d5f2245d9fa83940abcc114a84d3f#diff-64070d7a771b2a8229f1551983f12235d058b6af2fce4a834a626fd13eced843L566), you removed use_iscrowd from the public API of DetectionDataset.from_coco. I agree with this change. However, I think we should keep the use_iscrowd parameter available at the load_coco_annotations level.

SkalskiP commented on supervision/detection/utils.py:
```diff
@@ -1325,3 +1325,77 @@ def spread_out_boxes(
+    return Ai / (Aa + Ab - Ai + EPS)
+
+
+def iou_with_jaccard(
```
From your explanation over email, I understand that we cannot merge this function with box_iou_batch. In that case, let’s rename it to box_iou_batch_with_jaccard and add it to supervision/__init__.py.

rafaelpadilla replied:
Got your idea now! My bad 😅
I was a bit too literal with your previous comments.
:heavy_check_mark: Fixed now.

SkalskiP commented on supervision/detection/utils.py:
```diff
@@ -1325,3 +1325,77 @@ def spread_out_boxes(
+
+    Returns:
+        np.ndarray: Array of IoU values of shape (len(dt), len(gt)).
+    """
```
Please include usage example at the end of docstring. You can use xyxy_to_xywh as reference.

rafaelpadilla replied:
:heavy_check_mark: Done!

rafaelpadilla replied:
:heavy_check_mark: Done!

SkalskiP commented:
Hi @rafaelpadilla,
One of the features currently missing from our mAP implementation is support for max_det. The original pycocotools implementation allows you to specify the maximum number of detections per image and automatically sorts and filters them accordingly.

rafaelpadilla commented:
Hi @SkalskiP ,

I am not aware of this parameter `max_det` in the official pycocotools. Actually I have never seen the limitation of max detections by the user.  

I was looking for this usage in their examples provided by the official page [here](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocoEvalDemo.ipynb) and [here](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocoDemo.ipynb) , and could not find it. Where is it passed to the evaluator?

Could you please provide an example of a pycocotools call using this `max_det` parameter?

SkalskiP commented:
@rafaelpadilla 

> I am not aware of this parameter `max_det` in the official pycocotools. Actually I have never seen the limitation of max detections by the user.

I actually meant `maxDet`. I used snake_case out of habit, so I wrote `max_det`. Sorry for the confusion!

> If you are referring to `maxDet` in [this](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py#L235) function, this is not an argument passed by the user. It is actually used internally for the computation of mAP. 

I genuinely thought that `maxDet` was a configurable parameter. In that case, let's keep the functional consistency with `pycocotools`. Please ignore my previous comment.

SkalskiP commented on supervision/dataset/formats/coco.py:
```diff
@@ -159,16 +173,29 @@ def detections_to_coco_annotations(
     images_directory_path: str,
     annotations_path: str,
     force_masks: bool = False,
+    use_iscrowd: bool = True,
```
I'ts a bit weird that by default `use_iscrowd=True` in `load_coco_annotations` but `use_iscrowd=False` in `coco_annotations_to_detections`. Let's unify that.

rafaelpadilla replied:
Agree! :+1: 
:heavy_check_mark: Unified to `use_iscrowd=True`

SkalskiP commented on supervision/dataset/formats/coco.py:
```diff
@@ -159,16 +173,29 @@ def detections_to_coco_annotations(
     return coco_annotations, annotation_id
 
 
+def get_coco_class_index_mapping(annotations_path: str) -> Dict[int, int]:
```
This function will become part of the public supervision API.

- Just like with box_iou_batch_with_jaccard, we need to IMPORT it in `supervision/__init__.py` and include it in `__all__`.

- We also need a docstring that explains the purpose of this function, the use cases where it would be useful, and how it works.

rafaelpadilla replied:
:heavy_check_mark: Done!

rafaelpadilla commented on supervision/__init__.py:
```diff
@@ -229,3 +233,103 @@
 ]
+
+
+def _jaccard(box_a: List[float], box_b: List[float], is_crowd: bool) -> float:
```
:heavy_check_mark: Done!

rafaelpadilla commented on supervision/dataset/formats/coco.py:
```diff
@@ -90,7 +90,10 @@ def coco_annotations_to_masks(
 
 
 def coco_annotations_to_detections(
```
Nice!
:heavy_check_mark:  Done!

For every existing test case in `test_coco_annotations_to_detections`, now we have test cases with `use_iscrowd=True` and `use_iscrowd=False`.

rafaelpadilla commented:
Hi @SkalskiP 

I believe I have addressed all the points raised during the review. 🙌
I also expanded the test cases to improve coverage with use_iscrowd=True and use_iscrowd=False.

In this last round of updates, nothing was changed that would affect the outputs. Still, as an extra precaution, I re-ran everything to confirm that the results remain consistent with previous runs ✅

The pre-commit.ci check shows a failure that is not related to this PR.

Let me know if everything looks good on your end. I believe this is now in a solid state to be merged. 🚀
Hopefully, this update will help the community achieve more consistent and reliable results with supervision.

Thanks again for the thoughtful feedback throughout the review process! 🙏 🙂
